main.cc

```diff
@@ -380,10 +380,15 @@ class Peer : public SocketHandler {
 
 		// Some error on the socket.
 		state_ = STATE_TCP_CONNECTING;
-		if (errno != EINPROGRESS && errno != EAGAIN) {
-			if (!stat.tcp_connections)
-				throw Except("cannot establish even one"
-					     " TCP connection");
+		if ((errno != EINPROGRESS && errno != EAGAIN) || ret) {
+			if (!stat.tcp_connections) {
+				std::stringstream ss;
+				ss << "cannot establish even one TCP connection";
+				if (ret)
+					ss << " (" << strerror(ret)
+					   << ", errno=" << errno << ")";
```

Comment:
`Except` already handles `errno` for `errno != EINPROGRESS && errno != EAGAIN`, doesn't it? If so, then why to do this here? Anyway, the string forming with `errno` should be done inside `Except` - there is no point to assemble the strings each time when we need to throw an exception